implement tuning values for elasticity

AssassinTee opened the pull request:
implements `ground_elasticity_x` and `ground_elasticity_y`

Similar to https://github.com/teeworlds/teeworlds/pull/3136

~closes #5124~ implements the player bouncing part of this. Doesn't effect projectiles!

can be followed up by #5313 with a "slimey block", see [here](https://github.com/ddnet/ddnet/issues/5124#issuecomment-1140402547)



## Checklist

- [x] Tested the change ingame
- [ ] ~Provided screenshots if it is a visual change~ NO VISUALS
- [x] Tested in combination with possibly related configuration options
- [ ] Written a unit test if it works standalone, system.c especially
- [x] Considered possible null pointers and out of bounds array indexing
- [x] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

def- commented:
Cool, looks like we're getting fancy new physics :)

C0D3D3V commented:
new tunings should be added to the end of the list, or you fix the code weher we use hard coded tuning indexes.

C0D3D3V commented:
I think we should also add a version check in CGameContext::SendTuningParams (similar to the other version checks).
Deen will hopefully clarify what the next version will be :)

AssassinTee commented:
> I think we should also add a version check in CGameContext::SendTuningParams (similar to the other version checks). Deen will hopefully clarify what the next version will be :) (or you break out of the pattern and check for the current version)

Can you elaborate? I would set a paramter like `VERSION_DDNET_ELASTICITY_TUNE = 17000` (to a yet not existing new version of DDNet)

C0D3D3V commented:
@def- Do you want to release 16.2 or jump to 17? I would be fine with 16.2, to consider is that we have some major changes like the changed input handling... 

@AssassinTee  yes that would be the way... the question is only if you should use 16020 or 17000

AssassinTee commented:
@C0D3D3V older clients can't handle elasticity at all, they would get prediction errors. I suggest doing a major break to 17000 for this

AssassinTee commented:
@C0D3D3V do I need to upgrade the Client Version in this PR, too?

C0D3D3V commented:
No deen releases new versions.

def- commented:
bors r+

heinrich5991 commented:
bors r-

C0D3D3V commented:
@heinrich5991 I forgot why we canceled the merge. Do you remember? What is missing?

AssassinTee commented:
@C0D3D3V @heinrich5991 I noticed one problem with this but this may be out of scope: you don't always get your doublejump back, this is inconsistent! The reason is that you are only touching the ground in a subtick and not in the tick before or after the touch. I would say, that this is out of scope for this issue and also changing this would change physics.

The workaround would be adding a flag to MoveBox and setting the jump values accordingly

C0D3D3V commented:
If a mapper would like to force refill of dj, he could add a jump refill till.

heinrich5991 commented:
Ah, I'd actually like to make this in-scope, so that we don't have a physics change due to this later on.

Chairn commented:
Fine by me.

C0D3D3V commented:
Do we even want that the dj refills? I guess that way you could increase your jump hight with every bounce? I did not test yet the physic...

But on the other hand, if you do not want to refill the jump you could also give the player 0 jumps

heinrich5991 commented:
> Do we even want that the dj refills? I guess that way you could increase your jump hight with every bounce? I did not test yet the physic...

If we don't want them, it should be consistent that we don't get it.

heinrich5991 commented:
Status: There's still an unaddressed review comment.

AssassinTee commented:
> Status: There's still an unaddressed review comment.

I somehow missed this reviews, I already got the code for this, it just needs to be integrated! :)
One side note: The sub-tick refill is only for slime blocks, since you can touch normal blocks in a subtick between ticks without getting your jump back and this is standard behaviour. In race we call this an edge-bug. My goal is not changing default physics at all

heinrich5991 commented:
Status: There's still an unaddressed review comment. Do you still want to work on this? @AssassinTee

AssassinTee commented:
> If we don't want them, it should be consistent that we don't get it.

I discussed this issue with heinrich a bit more in detail in discord, but here are the keynotes:
You get the jump back based on ticks, but you can skip a ground-touch in a subtick by bouncing off of the block between two ticks, so I added a check, if you hit the ground in between in order to give the jump back (default behavior). You can't accelerate in stupid ways with it, because if you jump your speed is set to the ground or airjump speed (it's not additive).

![ElasticityGraphic](https://github.com/ddnet/ddnet/assets/821401/158862c7-62c8-4718-afa5-84dc165cd391)

Why do you only give the jump back, if the elasiticy > 0? Because I want to keep the subtick bug that is present in the current version.

heinrich5991 commented on src/game/collision.cpp:
```diff
@@ -513,14 +516,16 @@ void CCollision::MoveBox(vec2 *pInoutPos, vec2 *pInoutVel, vec2 Size, float Elas
+					Vel.y *= -ElasticityY;
 					Hits++;
+					if(pGrounded && ElasticityY > 0)
+						*pGrounded = true;
```
Do you also get your DJ back when hitting the ceiling? Looks like that to me. Would be undesired.

AssassinTee replied:
Fixed in the new version, thanks for the input!

heinrich5991 commented:
Thanks for your persistence!
